### ISSUE TITLE
Use inference_mode instead of no_grad for pth v1.9.0

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/ignite#2193]
## 🚀 Feature

The idea is to replace where it is appropriate `torch.no_grad` with `inference_mode` to speed-up computations: evalutation, metrics.

This works since pytorch v1.9.0.

- https://pytorch.org/docs/1.9.0/notes/autograd.html#inference-mode
- https://pytorch.org/docs/1.9.0/generated/torch.inference_mode.html?highlight=inference_mode#torch.inference_mode

Let's also benchmark the speed-up on a simple case.

**Comment by KickItLikeShika** [comment]
I will work on this, `torch.no_grad` exists in:
- Examples
- Docs 
- Metrics
- Engine
I will start workin on them in this same order, what do you think?

**Comment by vfdev-5** [comment]
@KickItLikeShika it is more a discussion issue. Can you first benchmark how quicker `inference_mode` vs `no_grad` for one of our examples, e.g. cifar10

**Comment by KickItLikeShika** [comment]
Cool, will do that on CIFAR and share the results, maybe more complex use case like CIFAR, can get us more clear benchmarks.

**Comment by vfdev-5** [comment]
I agree. To see more clearly we can just benchmark metric updates

**Comment by KickItLikeShika** [comment]
Ran CIFAR10 example for 5 epochs each (no_grad, inference_mode), and averaged the time for evaluating test metrics. (on Colab GPU).
- torch.no_grad: 3.7 sec
- torch.inference_mode: 3.688 sec
NB: https://colab.research.google.com/drive/1zfWxg8H0XOutqXfh8BORXaz2sJ4hFE_-?usp=sharing

**Comment by vfdev-5** [comment]
The benchmark in pytorch does not show any benefit of changing: https://github.com/pytorch/ignite/pull/3646#issuecomment-4010711729

Closing as non planned
